spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/FirestoreTemplate.java

```diff
@@ -187,7 +241,32 @@ public FirestoreTemplate(FirestoreStub firestore, String parent) {
 		FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
 		Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
 
-		return this.parent + "/" + persistentEntity.collectionName() + "/" + idVal.toString();
+		return buildResourceName(persistentEntity, idVal.toString());
+	}
+
+	private String buildResourceName(FirestorePersistentEntity<?> persistentEntity, String s) {
+		return this.parent + "/" + persistentEntity.collectionName() + "/" + s;
+	}
+
+	public String getIdValue(Object entity, FirestorePersistentEntity persistentEntity) {
+		FirestorePersistentProperty idProperty = persistentEntity.getIdPropertyOrFail();
+		Object idVal = persistentEntity.getPropertyAccessor(entity).getProperty(idProperty);
+
+		return idVal.toString();
+	}
+
+	static private class DocumentReadContext {
```

Comment:
If you take my above suggestion about returning `Mono<Document>` from getDocument, then it will allow you to remove this class.